### DISCUSSION TITLE
Where to save (uploaded) files without loosing them on deploy

### DISCUSSION THREAD

**Discussion by dspangenberg** [adonisjs/core#1552] [Help]
Hi,

`Application.tmpPath()` or `Application.publicPath()` return the directories under 'build'. But these directories will be deleted after `node ace build --production`. So where to save the files without losing then after a deploy? And how to serve it later (http://127.0.0.1/media/example.jpeg)?

Thanks in advance and best regards,
Danny

**Comment by prashanth1k** [comment]
Build is transient - so those folders will be deleted :). You could version control build folders if need be, or create a distinct build repo to maintain build files in production. Depending on your use case that may be easier to revert (?), subject to additional build scripts (e.g. revert DB, or changes to files) and arguably cleaner approach.

**Reply by ashokgelal** [reply]
I don't think @dspangenberg is asking to save the build files but all the user uploaded files like profile pictures, documents etc. This is similar to the `storage` folder in Laravel. I recommend using a folder outside the build path (say: `/home/dspangenberg/your-site/media)` and then linking it after the deployment as a hook.

You can use a deployment tool like [Cleaver](https://cleaver.cloud), which supports Adonis deployments out of the box, to make this painless for you, like adding a hook after latest deployment activation. (Disclaimer: I'm the Co-Founder of Cleaver)

**Comment by thetutlage** [comment]
I recommend saving them outside of the `build` folder. Simply create the path as follows:

```ts
Application.makePath('../storage/uploads')
```

The `makePath` method makes a path from the app root (which in `build`)

**Comment by LeadcodeDev** [comment]
After doing @thetutlage advises you,  to retrieve the path of your file, you must create a route to access it :
```ts
Route.get('uploads/services/:filename', async ({ response, params }) => {
    const path = 'path_of_your_upload_folder'
    response.download(Application.makePath(path, params.filename))
})
```
